### PR TITLE
[Serializer] Reduce complexity of NameConverter

### PR DESCRIPTION
**Pull request by ostrolucky**
Cleaner and faster implementation of camelcase normalization.
Speed improvement is particularly visible in long string input.

| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

### CODE REVIEW

File: src/Symfony/Component/Serializer/NameConverter/CamelCaseToSnakeCaseNameConverter.php [hunk 1]
```diff
@@ -44,19 +44,7 @@ public function __construct(array $attributes = null, $lowerCamelCase = true)
...
-            for ($i = 0; $i < $len; ++$i) {
-                if (ctype_upper($lcPropertyName[$i])) {
-                    $snakeCasedName .= '_'.strtolower($lcPropertyName[$i]);
-                } else {
-                    $snakeCasedName .= strtolower($lcPropertyName[$i]);
-                }
-            }
-
-            return $snakeCasedName;
+            return strtolower(preg_replace('/[A-Z]/', '_\\0', lcfirst($propertyName)));
```
Comment (dosten) [inline src/Symfony/Component/Serializer/NameConverter/CamelCaseToSnakeCaseNameConverter.php]: Why do you need the `strtolower` call?

Comment (gadelat) [inline src/Symfony/Component/Serializer/NameConverter/CamelCaseToSnakeCaseNameConverter.php]: this preg_replace doesn't replace those uppercase characters with lowercase ones

**Comment by iltar** [comment]
Do you have before/after comparisons?

**Comment by robfrawley** [comment]
I'd be interested in before/after numbers as well.

**Comment by gadelat** [comment]
testing command: `time php ./vendor/bin/phpunit Tests/NameConverter --filter testNormalize --repeat 5000`. Results are based on 10 sample average.

Default testcase:
<table>
  <tr>
    <th>Implementation:</th>
    <td>loop</td>
   <td>regex</td>
  </tr>
  <tr>
    <td></td>
    <td>12,646s</td>
    <td>12,352s</td>
  </tr>
  <tr>
    <td colspan="3" align="center">2,3% improvement</td>
  </tr>
</table>

However, when you change testcase like this:

```diff
--- Tests/NameConverter/CamelCaseToSnakeCaseNameConverterTest.php	(revision b14a5f3efa9b6f24b55d022216e8c2e93e861459)
+++ Tests/NameConverter/CamelCaseToSnakeCaseNameConverterTest.php	(revision )
@@ -46,6 +46,7 @@
     public function attributeProvider()
     {
         return array(
+            array(str_repeat('coop_tilleuls', 100), str_repeat('coopTilleuls', 100), true),
             array('coop_tilleuls', 'coopTilleuls', true),
             array('_kevin_dunglas', '_kevinDunglas', true),
             array('this_is_a_test', 'thisIsATest', true),
```
<table>
  <tr>
    <th>Implementation:</th>
    <td>loop</td>
   <td>regex</td>
  </tr>
  <tr>
    <td></td>
    <td>19,955s</td>
    <td>13,626s</td>
  </tr>
  <tr>
    <td colspan="3" align="center">46,5% improvement</td>
  </tr>
</table>

Now, I'm not going to continue doing this with 10 samples and 5000 repeats, but if you change str_repeat argument from 100 to 10000, for 100 repeats regex is **25x** faster

**Comment by iltar** [comment]
So it's primarily a gain for bigger strings? Note that those are less common. However, it seems that even with the smallest strings it gains speed, so that's all good!

**Comment by gadelat** [comment]
Yep. What's most important is that this is more clean and doesn't have negative performance impact.

**Review by nicolas-grekas — APPROVED** [review]
:+1: (could be merged on a lower branch as bug fix to reduce potential future conflicts?)

**Comment by fabpot** [comment]
Thank you @gadelat.

File: src/Symfony/Component/Serializer/NameConverter/CamelCaseToSnakeCaseNameConverter.php (hunk 1 above)
Comment (rybakit) [inline src/Symfony/Component/Serializer/NameConverter/CamelCaseToSnakeCaseNameConverter.php]: An alternative version w/o `lcfirst()`:

```php
return strtolower(preg_replace('/(?<=[a-z])([A-Z])/', '_$1', $propertyName));
```

Comment (gadelat) [inline src/Symfony/Component/Serializer/NameConverter/CamelCaseToSnakeCaseNameConverter.php]: Doesn't pass the tests, is slowest of all and less readable

Comment (rybakit) [inline src/Symfony/Component/Serializer/NameConverter/CamelCaseToSnakeCaseNameConverter.php]: Indeed, the correct regex is `'/(?<=[A-Za-z])([A-Z])/'`, but it's still slower than lcfirst version.

**Pull request merged**